Add a check to ensure integrity of cache to gitClientFactory

mpherman2 opened the pull request:
/assign @cjwagner @listx

listx reviewed (commented):
I wonder if we can do all of the fsck on startup, instead of doing it lazily?

listx commented:
Ugh, GitHub deleted my long, carefully worded comment because it was on a deleted commit.

Anyway, to expand on my earlier comment above, if you make the fsck happen *before* a service (gerrit, sub, tide) starts up, then you don't have to worry about re-cloning on an fsck failure (you'd just delete the directory and that would be it).

mpherman2 commented:
Addressed the comments. With regard to when to run fsck. It is not the fastest command to run so I figure lazily executing it instead of running it a bunch on startup might help avoid a smaller coldstart issue. Especially since not all repos are going to need it at once.

listx commented:
> Addressed the comments. With regard to when to run fsck. It is not the fastest command to run so I figure lazily executing it instead of running it a bunch on startup might help avoid a smaller coldstart issue. Especially since not all repos are going to need it at once.

Wouldn't this be a non-issue, because the service would not serve its readiness endpoint until after the fsck is completed?

petr-muller commented on prow/git/v2/client_factory.go:
```diff
@@ -308,6 +315,33 @@ func (c *clientFactory) ClientFor(org, repo string) (RepoClient, error) {
+// Ensures that the repos in the cache are valid, clean, and up to date
+func (c *clientFactory) ensureValidUpdatedCache(cacheDir string, repoClient RepoClient, cacheClientCacher cacher) error {
+	// We only need to verify that the repos are valid once on startup
+	if _, ok := c.verifiedRepos[cacheDir]; !ok {
```
nit: maybe the function would read better (less indent and cache-like semantics ("if the repo is valid then just update it") if we checked for the opposite?

mpherman2 replied:
Done

petr-muller commented on prow/git/v2/client_factory.go:
```diff
@@ -207,6 +210,16 @@ type clientFactory struct {
+	if err := os.MkdirAll(cacheDir, os.ModePerm); err != nil && !os.IsExist(err) {
+		return err
+	}
+	if err := cacheClientCacher.MirrorClone(); err != nil {
+		return err
+	}
+	return nil
```
nit: could be just `return ccc.MirrorClone()`

mpherman2 replied:
Done

petr-muller commented on prow/git/v2/client_factory.go:
```diff
@@ -169,6 +169,7 @@ func NewClientFactory(opts ...ClientFactoryOpt) (ClientFactory, error) {
 		repoLocks:      map[string]*sync.Mutex{},
 		logger:         logrus.WithField("client", "git"),
 		cookieFilePath: o.CookieFilePath,
+		verifiedRepos:  map[string]bool{},
```
nit: we could use `sets.NewString()` (which is internally the same thing), Prow uses that a lot

mpherman2 replied:
Done

petr-muller reviewed (commented):
I do not have a strong opinion on startup vs lazy but lean to agreeing with Linus that a readiness should protect us enough so that slower startup should not be a problem and is probably expected a bit more than a latency on a random request that trips the recheck.

Left some comments inline, most of them are suggestions rather than requirements.

mpherman2 commented on prow/git/v2/client_factory.go:
```diff
@@ -308,6 +315,33 @@ func (c *clientFactory) ClientFor(org, repo string) (RepoClient, error) {
+	// We only need to verify that the repos are valid once on startup
+	if _, ok := c.verifiedRepos[cacheDir]; !ok {
+		// Ensure it is valid
+		if valid, _ := repoClient.Fsck(); !valid {
```
fsck needs to return an error because other implementations of the gitClientFactory need to return an error (namely the adapter needs to return a `notImplemented` error) For the implementation nil error will always also return true and non nil error will always return false.

mpherman2 reviewed (commented):
After implementing the non-lazy version of this, I realized that it would cause a race. What I want to avoid is a race where
1) New Gerrit instance runs FSCK
2) Old Gerrit instance puts a lock on a repo that has already been checked
3) New Gerrit instance is ready and begins serving traffic
4) Old Gerrit instance is shut down before removing lock on repo

This would make the fsck worthless as the old gerrit instance could still mess with the directory after the fsck occurs

listx commented:
/lgtm
/hold
Thanks @mpherman2! Unhold at your discretion.

mpherman2 commented:
/unhold
